core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
     Option(allPartitions.get(topicPartition)).getOrElse(HostedPartition.None)
   }
 
-  def isAddingReplica(topicPartition: TopicPartition, replicaId: Int): Boolean = {
+  def isAddingReplica(topicPartition: TopicPartition, replicaId: Int, includeFenced: Boolean): Boolean = {
     getPartition(topicPartition) match {
-      case Online(partition) => partition.isAddingReplica(replicaId)
+      case HostedPartition.Online(partition) => partition.isAddingReplica(replicaId)
+      case HostedPartition.Fenced(metadata) => includeFenced && metadata.partition.isAddingReplica(replicaId)
       case _ => false
     }
   }
 
   // Visible for testing
-  def createPartition(topicPartition: TopicPartition): Partition = {
+  def createPartition(topicPartition: TopicPartition, fenced: Boolean = false,
+                      metadataPartitionForFence: Option[MetadataPartition] = None,
+                      wasNewForFence: Option[Boolean] = None,
+                      mostRecentMetadataOffsetForFence: Option[Long] = None): Partition = {
     val partition = Partition(topicPartition, time, this)
-    allPartitions.put(topicPartition, HostedPartition.Online(partition))
+    allPartitions.put(topicPartition,
+      if (fenced) {
+        if (metadataPartitionForFence.isEmpty || wasNewForFence.isEmpty || mostRecentMetadataOffsetForFence.isEmpty)
+          throw new IllegalArgumentException(s"Must supply all metadata when creating a fenced partition: $topicPartition")
+        HostedPartition.Fenced(FencedPartitionMetadata(partition,
+          metadataPartitionForFence.get, wasNewForFence.get, mostRecentMetadataOffsetForFence.get))
+      } else {
+        HostedPartition.Online(partition)
+      })
     partition
   }
 
   def nonOfflinePartition(topicPartition: TopicPartition): Option[Partition] = {
     getPartition(topicPartition) match {
       case HostedPartition.Online(partition) => Some(partition)
+      case HostedPartition.Fenced(metadata) => Some(metadata.partition)
       case HostedPartition.None | HostedPartition.Offline => None
     }
   }
 
+  // not sure if any of the 9 invocations of the above nonOfflinePartition() should be diverted here
```

Comment:
Right. I don't think it's safe to expose the `Partition` object while it is in a fenced state. For example, `GroupMetadataManager` calls `Partition.appendRecordsToLeader` on the partition returned from `nonOfflinePartition`. But that would not be safe until the log has been initialized.

Reply:
> would not be safe until the log has been initialized

What is unsafe about it?  Looking at the code it throws `NotLeaderOrFollowerException` if the log doesn't exist, and if it does I assume it would be okay to append under the assumption that anything done incorrectly would not be committed.  Is this incorrect?

Reply:
We now scope all "deferred" methods as private within ReplicaManager.